### PR TITLE
quant aware with infer model

### PR DESCRIPTION
**Pull request by itminner**
imagenet 分类模型验证：
MobileNetV2: float模型[0.71918 0.90568]，量化模型[0.71764 0.90418]
MobileNetV3_small_x1_0: float模型[0.68    0.87896]，量化模型[0.6703 0.869]

### CODE REVIEW

File: demo/quant/quant_aware_with_infermodel/README.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+```
+
+- ``model_path``: 需要量化的模型所在路径
+- ``checkpoint_path``: 量化训练模型checkpoint保存路径
+- ``model_filename``: 如果需要量化的模型的参数文件保存在一个文件中，则设置为该模型的模型文件名称，如果参数文件保存在多个文件中，则不需要设置。
+- ``params_filename``: 如果需要量化的模型的参数文件保存在一个文件中，则设置为该模型的参数文件名称，如果参数文件保存在多个文件中，则不需要设置。
+
+运行以上命令后，可在``${checkpoint_path}``下看到量化后模型的checkpoint。
+
+## 量化模型导出
```
Comment (ceci3) [inline demo/quant/quant_aware_with_infermodel/README.md]: 这里为什么还需要一个单独的量化模型导出章节？

File: paddleslim/quant/quant_aware_with_infermodel.py [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+    assert isinstance(configs['teacher_model_path'], str), \
+        "'teacher_model_path' must both be float'"
+    assert isinstance(configs['model_path'], str), \
+        "'model_path' must both be str'"
+    return train_config
+
+
+def _create_optimizer(train_config):
+    """create optimizer"""
+    optimizer = paddle.optimizer.SGD(
```
Comment (ceci3) [inline paddleslim/quant/quant_aware_with_infermodel.py:107]: 优化器不需要支持其他优化器吗？ernie里用的一般都是AdamW

File: paddleslim/quant/quant_aware_with_infermodel.py [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+    "weight_decay": 0.0001,
+    "use_pact": False,
+    "quant_model_ckpt_path": "./quant_model_checkpoints/",
+    "teacher_model_path": None,
+    "teacher_model_filename": "__model__",
+    "teacher_params_filename": None,
+    "model_path": None,
+    "model_filename": "__model__",
+    "params_filename": None,
+    "distill_node_pair": None
```
Comment (ceci3) [inline paddleslim/quant/quant_aware_with_infermodel.py:62]: distill_node_pair 这个参数是以什么形式给出要蒸馏的节点？可以在文档里说明一下这个参数要怎么设置

File: paddleslim/quant/quant_aware_with_infermodel.py [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+    assert len(train_config["distill_node_pair"]) % 2 == 0, \
+        "distill_node_pair config wrong, the length needs to be an even number"
+    print("train config.distill_node_pair: ", train_config["distill_node_pair"])
+    distill_loss = 0
+    for i in range(len(train_config["distill_node_pair"]) // 2):
+        print(train_config["distill_node_pair"][i * 2],
+              train_config["distill_node_pair"][i * 2 + 1])
+        distill_loss += l2_loss(train_config["distill_node_pair"][i * 2],
+                                train_config["distill_node_pair"][i * 2 + 1])
+    print(distill_loss)
```
Comment (ceci3) [inline paddleslim/quant/quant_aware_with_infermodel.py]: 可以输出下打印的是什么内容

File: paddleslim/quant/quant_aware_with_infermodel.py [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+                    program=test_program,
+                    model_path=os.path.join(
+                        train_config["quant_model_ckpt_path"], checkpoint_name))
+            iter_sum += 1
+            if train_config["max_iter"] >= 0 and iter_sum > train_config[
+                    "max_iter"]:
+                return
+
+
+def export_quant_infermodel(
```
Comment (ceci3) [inline paddleslim/quant/quant_aware_with_infermodel.py:370]: 这个函数的具体作用是什么？

File: demo/quant/quant_aware_with_infermodel/README.md (hunk 1 above)
Comment (itminner) [reply demo/quant/quant_aware_with_infermodel/README.md]: 训练时，保存的是convert前的checkpoint，导出函数用于convert + save inference model

File: paddleslim/quant/quant_aware_with_infermodel.py (hunk 2 above)
Comment (itminner) [reply paddleslim/quant/quant_aware_with_infermodel.py:107]: 可以扩增，量化一般都是用最小级学习率慢慢调整，我理解SGD是够用的

File: paddleslim/quant/quant_aware_with_infermodel.py (hunk 3 above)
Comment (itminner) [reply paddleslim/quant/quant_aware_with_infermodel.py:62]: 嗯，我补充下文档

File: paddleslim/quant/quant_aware_with_infermodel.py (hunk 5 above)
Comment (itminner) [reply paddleslim/quant/quant_aware_with_infermodel.py:370]: 实现了convert + save inference model功能

File: demo/quant/quant_aware_with_infermodel/README.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+- ``'train'``文件夹，训练图片
+- ``'train_list.txt'``文件
+- ``'val'``文件夹，验证图片
+- ``'val_list.txt'``文件
+
+### 准备需要量化的模型
+
+本功能只支持加载通过``paddle.static.save_inference_model``接口保存的模型。因此如果您的模型是通过其他接口保存的，需要先将模型进行转化。本示例将以分类模型为例进行说明。
+
+首先在[imagenet分类模型](https://github.com/PaddlePaddle/models/tree/develop/PaddleCV/image_classification#%E5%B7%B2%E5%8F%91%E5%B8%83%E6%A8%A1%E5%9E%8B%E5%8F%8A%E5%85%B6%E6%80%A7%E8%83%BD)中下载训练好的``mobilenetv1``模型。
```
Comment (qingqing01) [inline demo/quant/quant_aware_with_infermodel/README.md]: 这里太老了，还是静态图的模型，已经很久不维护了。最好用paddleclas，或者paddle.vision.models内置的

File: demo/quant/quant_aware_with_infermodel/export_quantmodel.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
+import os
```
Comment (qingqing01) [inline demo/quant/quant_aware_with_infermodel/export_quantmodel.py:15]: add copyright

File: demo/quant/quant_aware_with_infermodel/quant_aware_with_infermodel.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
+import os
```
Comment (qingqing01) [inline demo/quant/quant_aware_with_infermodel/quant_aware_with_infermodel.py:15]: add copyright

File: demo/quant/quant_aware_with_infermodel/quant_aware_with_infermodel.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+DATA_DIR = "../../data/ILSVRC2012/"
+def eval(exe, place, compiled_test_program, test_feed_names, test_fetch_list):
+    val_reader = paddle.batch(reader.val(), batch_size=1)
+    image = paddle.static.data(
+        name='image', shape=[None, 3, 224, 224], dtype='float32')
+    label = paddle.static.data(name='label', shape=[None, 1], dtype='int64')
+
+    valid_loader = paddle.io.DataLoader.from_generator(
+        feed_list=[image], capacity=512, use_double_buffer=True, iterable=True)
+    valid_loader.set_sample_list_generator(val_reader, place)
```
Comment (qingqing01) [inline demo/quant/quant_aware_with_infermodel/quant_aware_with_infermodel.py]: 需要新的 DataLoader

File: paddleslim/quant/quant_aware_with_infermodel.py [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+
+DistillProgramInfo = namedtuple("DistillProgramInfo", \
+    "startup_program train_program train_feed_names train_fetch_list \
+     optimizer test_program test_feed_names test_fetch_list"
+                                                            )
+
+
+def build_distill_prog_with_infermodel(executor, place, train_config):
+    """build distill program with infermodel"""
+    [train_program, feed_target_names, fetch_targets]= fluid.io.load_inference_model( \
```
Comment (qingqing01) [inline paddleslim/quant/quant_aware_with_infermodel.py]: paddle.static. load_inference_model ， 另外需要确认，paddle.jit.load 接口是否可行

File: paddleslim/quant/quant_aware_with_infermodel.py [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+            initializer=paddle.nn.initializer.Constant(value=init_thres),
+            regularizer=paddle.regularizer.L2Decay(0.0001),
+            learning_rate=1)
+        u_param = helper.create_parameter(
+            attr=u_param_attr, shape=[1], dtype=dtype)
+
+        part_a = paddle.nn.functional.relu(x - u_param)
+        part_b = paddle.nn.functional.relu(-u_param - x)
+        x = x - part_a + part_b
+        return x
```
Comment (qingqing01) [inline paddleslim/quant/quant_aware_with_infermodel.py]: 不用实现在这个函数里，看下其他文件里是否有此函数实现，有的话，直接import即可

File: paddleslim/quant/quant_aware_with_infermodel.py (hunk 5 above)
Comment (qingqing01) [reply paddleslim/quant/quant_aware_with_infermodel.py:370]: 原先的export函数不能用吗？

File: tests/test_quant_aware_with_infermodel.py [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+                result[0].append(cost)
+                result[1].append(top1)
+                result[2].append(top5)
+            print(' avg loss {}, acc_top1 {}, acc_top5 {}'.format(
+                np.mean(result[0]), np.mean(result[1]), np.mean(result[2])))
+            return np.mean(result[1]), np.mean(result[2])
+
+        train(main_prog)
+        top1_1, top5_1 = test(val_prog)
+        paddle.fluid.io.save_inference_model(
```
Comment (qingqing01) [inline tests/test_quant_aware_with_infermodel.py]: 接口不规范

File: demo/quant/quant_aware_with_infermodel/README.md [hunk 13]
```diff
@@ -0,0 +1,121 @@
+# 使用infermodel进行量化训练示例
```
Comment (qingqing01) [inline demo/quant/quant_aware_with_infermodel/README.md]: 文档也需要规范， inferemodel 一般人不一定清楚是啥，连起来写也不规范

File: paddleslim/quant/quant_aware_with_infermodel.py (hunk 5 above)
Comment (itminner) [reply paddleslim/quant/quant_aware_with_infermodel.py:370]: 我们还没有在slim底层提供量化模型export为infermodel的函数。
这个函数可以实现只输入float infermodel 、量化配置和 checkpoint，实现量化训练checkpoint到infermodel转换，使用更加方便

File: demo/quant/quant_aware_with_infermodel/README.md (hunk 6 above)
Comment (itminner) [reply demo/quant/quant_aware_with_infermodel/README.md]: 好的

File: paddleslim/quant/quant_aware_with_infermodel.py (hunk 11 above)
Comment (itminner) [reply paddleslim/quant/quant_aware_with_infermodel.py]: 这个功能我们原来实现在了demo里，我看看怎么独立出来

File: paddleslim/quant/quant_aware_with_infermodel.py [hunk 14]
```diff
@@ -0,0 +1,507 @@
...
+        act_preprocess_func = None
+        optimizer_func = None
+        pact_executor = None
+
+    test_program = quant_aware(
+        test_program,
+        place,
+        quant_config,
+        scope=None,
+        act_preprocess_func=act_preprocess_func,
```
Comment (ceci3) [inline paddleslim/quant/quant_aware_with_infermodel.py:437]: convert应该不能支持量化方式为导出为pact的program，这个可以给默认的test program格式

Comment (itminner) [reply paddleslim/quant/quant_aware_with_infermodel.py:437]: 是的，导出模型不用pact

**Review by qingqing01 — APPROVED** [review]
